Add code example for each supported backend

w0xlt opened the pull request:
This PR adds code example for connecting to Esplora, Electrum Server, Neutrino and Bitcoin Core.
Also shows how to retrieve balance, sign and broadcast transactions.

To test:
```
cd examples/backend/
cargo run electrum
cargo run esplora
cargo run neutrino
cargo run rpc_core
```

rajarshimaitra reviewed (commented):
Concept ACK on adding more example to the project. But I think its possible without adding a new cargo project in the workspace. 

Can't we just have running examples like this https://github.com/bitcoindevkit/bdk/pull/519?

notmandatory commented:
Hi, please rebase to pickup changes in #596. Thanks!

w0xlt commented:
@notmandatory sorry for the delay.

I force pushed a new version that adds examples just for Electrum and Esplora, as there are already examples for Compact Filter and RPC backends.

I also updated the code to 0.20.1 and addressed @rajarshimaitra's suggestion not to add a new cargo project.

w0xlt commented:
The new push added the documentation suggested in https://github.com/bitcoindevkit/bdk/issues/685, explaining how to use `use-esplora-reqwest`.

w0xlt committed:
add `electrum_backend` example.

w0xlt committed:
add `esplora_backend` example.

rajarshimaitra commented on Cargo.toml:
```diff
@@ -116,6 +116,15 @@ name = "rpcwallet"
+[[example]]
+name = "electrum_backend"
+path = "examples/electrum_backend.rs"
+
```
You also need to pass the required feature as electrum here, or else all the tests with `--no-default-features` will try to compile this example and fail..

w0xlt replied:
Done in https://github.com/bitcoindevkit/bdk/pull/526/commits/8140d130edc065362ba636828c7a8ee5afb8f884

rajarshimaitra commented on examples/esplora_backend.rs:
```diff
@@ -0,0 +1,81 @@
+/// Users wanting asynchronous HTTP calls should change `required-features = ["use-esplora-ureq"]` to
+/// `required-features = ["use-esplora-reqwest"]` in `esplora_backend` target in Cargo.toml and run
+/// `cargo run --features="use-esplora-reqwest, reqwest-default-tls" --example esplora_backend`
+/// Without `reqwest-default-tls`, the asynchronous version will not work.
```
This is not fully correct.. In order to use async calls they also need to use the `async-interface` feature along with `use-esplora-reqwest`.. And when they do that this same example will not work anymore.. For example you need to then call `wallet.sync().await.unwrap()` ..

So just changing the required feature doesn't work. if we need to show case the async reqwest thing we need to create another example..

What I would suggest here is use the async and reqwest in the example instead of ureq.. This example is very similar to the electrum one, so just doing another blocking esplora doesn't add much value.. Instead we should do a non blocking reqwest to show how to use the async-interface of bdk..

w0xlt replied:
`wallet.sync().await.unwrap()` causes the error below:

```
error[E0277]: `Result<(), bdk::Error>` is not a future
  --> examples/esplora_backend.rs:52:53
   |
52 |     wallet.sync(&blockchain, SyncOptions::default()).await.unwrap();
   |                                                     ^^^^^^ `Result<(), bdk::Error>` is not a future
   |
   = help: the trait `std::future::Future` is not implemented for `Result<(), bdk::Error>`
   = note: Result<(), bdk::Error> must be a future or must implement `IntoFuture` to be awaited
   = note: required because of the requirements on the impl of `IntoFuture` for `Result<(), bdk::Error>`
help: remove the `.await`
   |
52 -     wallet.sync(&blockchain, SyncOptions::default()).await.unwrap();
52 +     wallet.sync(&blockchain, SyncOptions::default()).unwrap();
   | 
```

w0xlt replied:
If I run `wallet.sync(&blockchain, SyncOptions::default()).unwrap();` using `use-esplora-reqwest, reqwest-default-tls`, it runs without error, but the wallet shows empty balance because apparently it does not sync.

rajarshimaitra replied:
I think you also need to pass the `async-interface` feature.. Thats what makes the bdk reqwest codes async.. The error seems to be indicating that.. Its getting a regular `Result<>` whereas it expected a `Future`..

w0xlt replied:
I had removed `async-interface` due to error below.
It is weird because `esplora_backend` doesn't have any reference to `electrum` .

```bash
$ cargo run --features="use-esplora-reqwest, reqwest-default-tls, async-interface" --example esplora_backend

   Compiling bdk v0.20.1-dev (/home/node01/Dev/wbdk)
error: Features async-interface and electrum are mutually exclusive and cannot be enabled together
   --> src/lib.rs:214:1
    |
214 | / compile_error!(
215 | |     "Features async-interface and electrum are mutually exclusive and cannot be enabled together"
216 | | );
    | |_^
```

rajarshimaitra replied:
I think getting `reqwest` working is not always straight forward and might need some trial and errors.. I think these examples are still very useful without `reqwest` and with `ureq` instead.. That might help us move this one forward, and do the async example in a independent PR. 

How does that sound to you @w0xlt ??

w0xlt replied:
I think it is working now.
Apparently https://github.com/bitcoindevkit/bdk/pull/526/commits/400cbad4388322f4f0e0007365bf1a68ea95a6b0 is working with `reqwest`.

rajarshimaitra commented on examples/electrum_backend.rs:
```diff
@@ -0,0 +1,88 @@
+    run(&network, electrum_url, xpriv);
+}
+
+pub fn load_or_create_wallet(
```
This doesn't seem to be loading any wallet, and creating a new one everytime you run the example with memory database. So maybe just call it `create_wallet`??

w0xlt replied:
Done in https://github.com/bitcoindevkit/bdk/pull/526/commits/8140d130edc065362ba636828c7a8ee5afb8f884 and https://github.com/bitcoindevkit/bdk/pull/526/commits/400cbad4388322f4f0e0007365bf1a68ea95a6b0.

rajarshimaitra reviewed (changes requested):
Concept ACK.. Few more fixes needs to be done..

rajarshimaitra commented on examples/electrum_backend.rs:
```diff
@@ -0,0 +1,88 @@
+    // Apparently it works only with Electrs (not EletrumX)
+    let blockchain = ElectrumBlockchain::from(Client::new(electrum_url).unwrap());
+
+    let wallet = load_or_create_wallet(&network, &xpriv);
```
You need to update the function name here..

w0xlt replied:
My mistake. Done in 7a47022ed697e87cb3f978bdeb4aef0c90c39d33.

rajarshimaitra commented on examples/esplora_backend.rs:
```diff
@@ -0,0 +1,83 @@
+
+    let blockchain = EsploraBlockchain::new(esplora_url, 20);
+
+    let wallet = load_or_create_wallet(&network, &xpriv);
```
same here..

w0xlt replied:
My mistake. Done in a9c426e25d913d3fe4712a3a494ecf25081cfd18.

rajarshimaitra commented on examples/electrum_backend.rs:
```diff
@@ -0,0 +1,88 @@
+    let network = Network::Testnet;
+
+    let xpriv = "tprv8ZgxMBicQKsPcx5nBGsR63Pe8KnRUqmbJNENAfGftF3yuXoMMoVJJcYeUw5eVkm9WBPjWYt6HMWYJNesB5HaNVBaFc1M6dRjWSYnmewUMYy";
+        "health lyrics appear aunt either wrist maple hover family episode seven maze";
```
I did not understand why this mnemonic is here..

w0xlt replied:
I originally used mnemonic instead of xpriv and forgot to removed it. 
Removed in 7a47022ed697e87cb3f978bdeb4aef0c90c39d33.

rajarshimaitra commented on Cargo.toml:
```diff
@@ -116,6 +116,16 @@ name = "rpcwallet"
+[[example]]
+name = "esplora_backend"
+path = "examples/esplora_backend.rs"
+required-features = ["use-esplora-reqwest", "reqwest-default-tls"]
```
You also need to update here to make it ureq only..

w0xlt replied:
I didn't quite understand the suggestion. Could you elaborate more?

rajarshimaitra replied:
Sorry my bad.. I misunderstood the doc comment.. This is fine..

rajarshimaitra commented on examples/electrum_backend.rs:
```diff
@@ -0,0 +1,88 @@
+/// If enough amount is available, this will send a transaction to an address.
+/// Otherwise, this will display a wallet address to receive funds.
+///
+/// This can be run with `cargo run --features="electrum" --example electrum_backend` in the root folder.
```
nit: can omit the quote 
```suggestion
/// This can be run with `cargo run --features=electrum --example electrum_backend` in the root folder.
```

w0xlt replied:
Done in 7a47022ed697e87cb3f978bdeb4aef0c90c39d33.

rajarshimaitra commented on examples/esplora_backend.rs:
```diff
@@ -0,0 +1,83 @@
+/// The configuration above uses asynchronous HTTP calls.
+/// To run it in a synchronous way, the user should change `required-features` to
+/// `required-features = ["use-esplora-ureq"] in `esplora_backend` target in Cargo.toml and run
+/// `cargo run --features="use-esplora-ureq" --example esplora_backend`
```
Can we put this in a `Note` box or something.. Feels like the actual command to run it via reqwest is lost in the middle..

w0xlt replied:
Done in a9c426e25d913d3fe4712a3a494ecf25081cfd18.

rajarshimaitra reviewed (changes requested):
Code looks good.. It need few more fixes to work.. Sorry if I reviewed prematurely.

rajarshimaitra commented on examples/electrum_backend.rs:
```diff
@@ -0,0 +1,89 @@
+    run(&network, electrum_url, xpriv);
+}
+
+pub fn create_wallet(network: &Network, xpriv: &ExtendedPrivKey) -> Wallet<MemoryDatabase> {
```
Does these functions need to be public? This applies to all other such functions..

w0xlt replied:
Good catch. Done in f13d5e6bbb4d375037cd54cac694ca04003420c8 and aa5db56f8b2d42e4cf36b002a6e8afebdbd77d93.

rajarshimaitra reviewed (commented):
tACK 95e9b7f5fc0ee5aa6586217c8b8ff6400eab6853

Code loos good.. Thanks for the quick update.. One last comment..

rajarshimaitra commented on examples/electrum_backend.rs:
```diff
@@ -0,0 +1,89 @@
+
+    println!("balance: {}", balance);
+
+    if balance > 6500 {
```
balance isn't an integer anymore.. Thats why tests are failing..

w0xlt replied:
Rebased and changed the comparison.

rajarshimaitra reviewed (changes requested):
ACK f13d5e6bbb4d375037cd54cac694ca04003420c8

One last fix to pass the tests..

rajarshimaitra reviewed (approved):
tACK c7e1247777e3862e0e863e070bf3a7187bda8736

LGTM.. One suggestion maybe for a future PR.

rajarshimaitra commented on examples/esplora_backend.rs:
```diff
@@ -0,0 +1,93 @@
+/// This can be run with `cargo run --features="use-esplora-reqwest, reqwest-default-tls" --example esplora_backend`
+/// in the root folder.
+///
+/// Note: The configuration above uses asynchronous HTTP calls.
```
Sorry I have missed this before. Nothing major, feel free to ignore for this PR. 

Even though the calls are async in `reqwest` they are still blocking in our impl, using the `await_or_block` macro like this, https://github.com/bitcoindevkit/bdk/blob/06310f1dd06641c36cb1980ee34489e03eadf96b/src/blockchain/esplora/reqwest.rs#L126 

So to actually make the wallet calls async in the example you need to use the `async-interface` feature too.. In that case you will have `await` in your wallets calls in the example.. 

This PR is good to go on its own.. Maybe as a future PR an idea.

 - Transform this to a blocking example using `ureq`.
 - Create another fully `async` example using `reqwest`.

w0xlt replied:
That would be great. But, as mentioned above, the command `cargo run --features="use-esplora-reqwest, reqwest-default-tls, async-interface" --example esplora_backend` results in error.

I think the problem is here (`src/lib:210`)

```rust
#[cfg(all(feature = "async-interface", feature = "electrum"))]
compile_error!(
    "Features async-interface and electrum are mutually exclusive and cannot be enabled together"
);
```

Even there is no `electrum` in the features, it seems this is enabled by default. Is there a way to disable it ?

notmandatory replied:
You can disable the default features (which are "key-value-db", "electrum") by adding `--no-default-features`, eg:
```
cargo run --no-default-features --features=use-esplora-reqwest,reqwest-default-tls,async-interface --example esplora_backend
```

w0xlt replied:
@notmandatory this worked. Thanks.

@rajarshimaitra I created two files `examples/esplora_backend_synchronous.rs` and `examples/esplora_backend_asynchronous.rs` in a2b6a408f7589f2b9682181b7ab57447a28d6866.

w0xlt replied:
I don't know why `CI / Rust fmt (pull_request)` failed. I  ran `cargo fmt --all` before committing.

rajarshimaitra commented on examples/esplora_backend_asynchronous.rs:
```diff
@@ -0,0 +1,91 @@
+
+    let wallet = create_wallet(network, &xpriv);
+
+    await_or_block!(wallet.sync(&blockchain, SyncOptions::default())).unwrap();
```
We don't need `await_or_block` as we know its gonna be async in this case..
```suggestion
    wallet.sync(&blockchain, SyncOptions::default()).await.unwrap();
```

w0xlt replied:
Done in e50fcb6c53964435d8a70f6fe396bf60bb75f1e2.

rajarshimaitra commented on examples/esplora_backend_asynchronous.rs:
```diff
@@ -0,0 +1,91 @@
+
+use bdk_macros::await_or_block;
+
+/// This will create a wallet from an xpriv and get the balance by connecting to an Esplora server.
```
Something like this.. 
```suggestion
/// This will create a wallet from an xpriv and get the balance by connecting to an Esplora server, using non blocking asynchronous calls with `reqwest`. 
```

w0xlt replied:
Done in e50fcb6c53964435d8a70f6fe396bf60bb75f1e2.

rajarshimaitra commented on examples/esplora_backend_synchronous.rs:
```diff
@@ -0,0 +1,88 @@
+
+use crate::utils::tx::build_signed_tx;
+
+/// This will create a wallet from an xpriv and get the balance by connecting to an Esplora server.
```
Similar suggestion. The idea is to explain what this example is for in the first line. 
```suggestion
/// This will create a wallet from an xpriv and get the balance by connecting to an Esplora server, using blocking calls with `ureq`.  
```

w0xlt replied:
Done in e50fcb6c53964435d8a70f6fe396bf60bb75f1e2.

rajarshimaitra commented on Cargo.toml:
```diff
@@ -119,6 +119,21 @@ name = "rpcwallet"
+[[example]]
+name = "esplora_backend_asynchronous"
+path = "examples/esplora_backend_asynchronous.rs"
+required-features = ["use-esplora-reqwest", "reqwest-default-tls"]
```
shouldn't this include `async-interface` too?

w0xlt replied:
Done in e50fcb6c53964435d8a70f6fe396bf60bb75f1e2.

rajarshimaitra commented on examples/electrum_backend.rs:
```diff
@@ -0,0 +1,87 @@
+/// If enough amount is available, this will send a transaction to an address.
+/// Otherwise, this will display a wallet address to receive funds.
+///
+/// This can be run with `cargo run --features=electrum --example electrum_backend` in the root folder.
```
This works with `cargo run --example electrum_backend` because `electrum` is there by default.

w0xlt replied:
Done in ea2b9f5a433f9b04a257e61a2a9ddd7f1526b324. Thanks for the review.

rajarshimaitra reviewed (changes requested):
ACK a2b6a408f7589f2b9682181b7ab57447a28d6866

Thanks for the split.. This now looks complete.. Few more nits..

rajarshimaitra commented:
> I don't know why CI / Rust fmt (pull_request) failed. I ran cargo fmt --all before committing.

It was some temporary problem with rust update server.. I restarted the tests and it passed..

rajarshimaitra reviewed (approved):
tACK e50fcb6c53964435d8a70f6fe396bf60bb75f1e2

w0xlt commented:
Rebased.
Is any additional action required ?

rajarshimaitra reviewed (approved):
Sorry took me some time to get back to this.. Thanks for pushing this through.. 

tACK f99a6b9f434e86a78695da72e9d28e6937c7e5db

All looks good to me, and great work separating the blocking and async example separate. This is good to go, no more changes should be required..

22f-cy reviewed (approved):
if [ "$PWD" = "$HOME" ]; then
